File: src/wrappers/csharp/Itscam.Sdk/JpegUtils.cs
// SPDX-License-Identifier: Proprietary
// Copyright (c) 2026 Pumatronix

using System;
using System.Collections.Generic;
using System.Text;

namespace Pumatronix.Itscam
{
    /// <summary>
    /// Utilities for extracting metadata from the JPEG COM marker (0xFF 0xFE)
    /// embedded by ITSCAM cameras.
    ///
    /// <para>
    /// The comment body uses a semicolon-delimited key=value format:
    /// <c>Placa=ABC1234;CoordPlaca=100x200,150x40;ClassifierList=[1,95,50,100,300,200]</c>
    /// </para>
    /// </summary>
    public static class JpegUtils
    {
        /// <summary>
        /// Extract the raw comment string from a JPEG's COM marker (0xFF 0xFE).
        /// Scans backwards from the end of the JPEG data for the COM marker.
        /// </summary>
        /// <param name="jpeg">Raw JPEG bytes.</param>
        /// <returns>The comment body, or <see cref="string.Empty"/> if not found.</returns>
        public static string ExtractComment(byte[] jpeg)
        {
            if (jpeg == null || jpeg.Length < 4)
                return string.Empty;

            if (jpeg[jpeg.Length - 1] != 0xD9)
                return string.Empty;

            for (int i = jpeg.Length - 1; i > 0; i--)
            {
                if (jpeg[i - 1] == 0xFF && jpeg[i] == 0xFE)
                {
                    int lenPos = i + 1;
                    if (lenPos > jpeg.Length - 2)
                        return string.Empty;

                    int commentLen = (jpeg[lenPos] << 8) | jpeg[lenPos + 1];
                    if (commentLen < 2)
                        return string.Empty;

                    int bodyPos = lenPos + 2;
                    if (bodyPos > jpeg.Length)
                        return string.Empty;

                    int bodyLen = commentLen - 2;
                    if (bodyLen > jpeg.Length - bodyPos)
                        return string.Empty;

                    return Encoding.ASCII.GetString(jpeg, bodyPos, bodyLen);
                }
            }

            return string.Empty;
        }

        /// <summary>
        /// Extract the raw comment string from a JPEG's COM marker (0xFF 0xFE).
        /// </summary>
        /// <param name="jpeg">Raw JPEG bytes.</param>
        /// <returns>The comment body, or <see cref="string.Empty"/> if not found.</returns>
        public static string ExtractComment(ReadOnlySpan<byte> jpeg)
        {
            if (jpeg.Length < 4)
                return string.Empty;

            if (jpeg[jpeg.Length - 1] != 0xD9)
                return string.Empty;

            for (int i = jpeg.Length - 1; i > 0; i--)
            {
                if (jpeg[i - 1] == 0xFF && jpeg[i] == 0xFE)
                {
                    int lenPos = i + 1;
                    if (lenPos > jpeg.Length - 2)
                        return string.Empty;

                    int commentLen = (jpeg[lenPos] << 8) | jpeg[lenPos + 1];
                    if (commentLen < 2)
                        return string.Empty;

                    int bodyPos = lenPos + 2;
                    if (bodyPos > jpeg.Length)
                        return string.Empty;

                    int bodyLen = commentLen - 2;
                    if (bodyLen > jpeg.Length - bodyPos)
                        return string.Empty;

                    return Encoding.ASCII.GetString(
                        jpeg.Slice(bodyPos, bodyLen).ToArray(), 0, bodyLen);
                }
            }

            return string.Empty;
        }

        /// <summary>
        /// Parse a JPEG comment string into a key=value dictionary.
        /// </summary>
        /// <param name="comment">Raw comment string.</param>
        /// <returns>Tag name to value mapping.</returns>
        public static IReadOnlyDictionary<string, string> ParseCommentTags(string comment)
        {
            var tags = new Dictionary<string, string>();
            if (string.IsNullOrEmpty(comment))
                return tags;

            foreach (var field in comment.Split(';'))
            {
                int eq = field.IndexOf('=');
                if (eq > 0)
                    tags[field.Substring(0, eq)] = field.Substring(eq + 1);
            }

            return tags;
        }

        /// <summary>
        /// One-shot convenience: extract the JPEG COM comment and parse it
        /// into a key=value tag dictionary.
        /// </summary>
        /// <param name="jpeg">Raw JPEG bytes.</param>
        /// <returns>Tag name to value mapping (empty if no COM marker).</returns>
        public static IReadOnlyDictionary<string, string> ParseCommentTags(byte[] jpeg)
        {
            return ParseCommentTags(ExtractComment(jpeg));
        }

        /// <summary>
        /// Extract structured plate-recognition results from parsed comment tags.
        /// </summary>
        public static IReadOnlyList<PlateRecognitionResult> ExtractPlateRecognitions(
            IReadOnlyDictionary<string, string> tags)
        {
            var plates = new List<PlateRecognitionResult>();
            if (!tags.TryGetValue("Placa", out var placa) || string.IsNullOrEmpty(placa))
                return plates;

            var plateTexts = placa.Split('_');
            var coordTexts = tags.TryGetValue("CoordPlaca", out var c) && !string.IsNullOrEmpty(c)
                ? c.Split('_') : Array.Empty<string>();
            var colorTexts = tags.TryGetValue("CorPlaca", out var col) && !string.IsNullOrEmpty(col)
                ? col.Split('_') : Array.Empty<string>();
            var countryTexts = tags.TryGetValue("OCRCountryCode", out var cc) && !string.IsNullOrEmpty(cc)
                ? cc.Split('_') : Array.Empty<string>();

            for (int i = 0; i < plateTexts.Length; i++)
            {
                if (string.IsNullOrEmpty(plateTexts[i]))
                    continue;

                var pr = new PlateRecognitionResult { Plate = plateTexts[i] };

                if (i < coordTexts.Length && !string.IsNullOrEmpty(coordTexts[i]))
                {
                    var parts = coordTexts[i].Split(',');
                    if (parts.Length >= 2)
                    {
                        var xy = parts[0].Split('x');
                        var wh = parts[1].Split('x');
                        if (xy.Length == 2)
                        {
                            int.TryParse(xy[0], out var px);
                            int.TryParse(xy[1], out var py);
                            pr.X = px; pr.Y = py;
                        }
                        if (wh.Length == 2)
                        {
                            int.TryParse(wh[0], out var pw);
                            int.TryParse(wh[1], out var ph);
                            pr.Width = pw; pr.Height = ph;
                        }
                    }
                }

                if (i < colorTexts.Length)
                    pr.Color = colorTexts[i];
                if (i < countryTexts.Length && int.TryParse(countryTexts[i], out var ccode))
                    pr.CountryCode = ccode;

                plates.Add(pr);
            }

            return plates;
        }

        /// <summary>
        /// Extract structured vehicle/object detection results from parsed
        /// comment tags.
        /// </summary>
        public static IReadOnlyList<ObjectDetectionResult> ExtractObjectDetections(
            IReadOnlyDictionary<string, string> tags)
        {
            var detections = new List<ObjectDetectionResult>();
            if (!tags.TryGetValue("ClassifierList", out var clsRaw) || string.IsNullOrEmpty(clsRaw))
                return detections;

            var clsEntries = ParseBracketEntries(clsRaw);
            var bmcEntries = tags.TryGetValue("BMCList", out var bmcRaw)
                ? ParseBracketEntries(bmcRaw)
                : new List<string>();

            for (int i = 0; i < clsEntries.Count; i++)
            {
                var parts = clsEntries[i].Split(',');
                if (parts.Length < 6) continue;

                if (!int.TryParse(parts[0], out var type) ||
                    !int.TryParse(parts[1], out var prob) ||
                    !int.TryParse(parts[2], out var x) ||
                    !int.TryParse(parts[3], out var y) ||
                    !int.TryParse(parts[4], out var w) ||
                    !int.TryParse(parts[5], out var h))
                    continue;

                var od = new ObjectDetectionResult
                {
                    Type = type,
                    Probability = prob / 100f,
                    X = x, Y = y, Width = w, Height = h
                };

                if (i < bmcEntries.Count)
                {
                    var cleaned = bmcEntries[i].Replace("\"", "");
                    var bmc = cleaned.Split(',');
                    if (bmc.Length >= 6)
                    {
                        od.Brand = bmc[0];
                        if (int.TryParse(bmc[1], out var bp)) od.BrandProb = bp / 100f;
                        od.Model = bmc[2];
                        if (int.TryParse(bmc[3], out var mp)) od.ModelProb = mp / 100f;
                        od.Color = bmc[4];
                        if (int.TryParse(bmc[5], out var cp)) od.ColorProb = cp / 100f;
                    }
                }

                detections.Add(od);
            }

            return detections;
        }

        private static List<string> ParseBracketEntries(string raw)
        {
            var entries = new List<string>();
            int pos = 0;
            while (pos < raw.Length)
            {
                int open = raw.IndexOf('[', pos);
                if (open < 0) break;
                int close = raw.IndexOf(']', open);
                if (close < 0) break;
                entries.Add(raw.Substring(open + 1, close - open - 1));
                pos = close + 1;
            }
            return entries;
        }
    }

    /// <summary>Plate recognition result parsed from JPEG COM metadata.</summary>
    public sealed class PlateRecognitionResult
    {
        public string Plate { get; set; } = string.Empty;
        public int X { get; set; }
        public int Y { get; set; }
        public int Width { get; set; }
        public int Height { get; set; }
        public string Color { get; set; } = string.Empty;
        public int CountryCode { get; set; }
    }

    /// <summary>Vehicle/object detection result parsed from JPEG COM metadata.</summary>
    public sealed class ObjectDetectionResult
    {
        public int Type { get; set; }
        public float Probability { get; set; }
        public int X { get; set; }
        public int Y { get; set; }
        public int Width { get; set; }
        public int Height { get; set; }
        public string Brand { get; set; } = string.Empty;
        public float BrandProb { get; set; }
        public string Model { get; set; } = string.Empty;
        public float ModelProb { get; set; }
        public string Color { get; set; } = string.Empty;
        public float ColorProb { get; set; }
    }
}

File: src/wrappers/nodejs/src/jpeg-utils.js
/*
 * SPDX-License-Identifier: Proprietary
 * Copyright (c) 2026 Pumatronix
 *
 * Pure-JavaScript implementation of the ITSCAM JPEG COM marker parser.
 * Mirrors src/wrappers/python/itscam/jpeg_utils.py.
 */
'use strict';

/** Extract the raw text of the JPEG COM marker (0xFF 0xFE), or "". */
function extractJpegComment(jpeg) {
    if (!jpeg || jpeg.length < 4) return '';
    const buf = Buffer.isBuffer(jpeg) ? jpeg : Buffer.from(jpeg);
    if (buf[buf.length - 1] !== 0xD9) return '';

    for (let i = buf.length - 1; i > 0; i--) {
        if (buf[i - 1] === 0xFF && buf[i] === 0xFE) {
            const lenPos = i + 1;
            if (lenPos > buf.length - 2) return '';
            const commentLen = (buf[lenPos] << 8) | buf[lenPos + 1];
            if (commentLen < 2) return '';
            const bodyPos = lenPos + 2;
            const bodyLen = commentLen - 2;
            if (bodyPos + bodyLen > buf.length) return '';
            return buf.slice(bodyPos, bodyPos + bodyLen)
                .toString('ascii');
        }
    }
    return '';
}

/** Split a `Key=Value;Key=Value` string into a plain object. */
function parseJpegCommentTags(comment) {
    const out = {};
    if (!comment) return out;
    for (const field of comment.split(';')) {
        const eq = field.indexOf('=');
        if (eq > 0) out[field.slice(0, eq)] = field.slice(eq + 1);
    }
    return out;
}

function _parseInt(s, fallback = 0) {
    if (s == null) return fallback;
    const n = parseInt(String(s).trim(), 10);
    return Number.isFinite(n) ? n : fallback;
}

function extractPlateRecognitions(tags) {
    const placa = tags && tags.Placa ? tags.Placa : '';
    if (!placa) return [];

    const plateTexts = placa.split('_');
    const coords = (tags.CoordPlaca || '').split('_');
    const colors = (tags.CorPlaca || '').split('_');
    const cc     = (tags.OCRCountryCode || '').split('_');

    const plates = [];
    for (let i = 0; i < plateTexts.length; i++) {
        const text = plateTexts[i];
        if (!text) continue;
        const pr = {
            plate: text, x: 0, y: 0, width: 0, height: 0,
            color: '', countryCode: 0,
        };
        if (i < coords.length && coords[i]) {
            const parts = coords[i].split(',');
            if (parts.length >= 2) {
                const xy = parts[0].split('x');
                const wh = parts[1].split('x');
                if (xy.length === 2) {
                    pr.x = _parseInt(xy[0]);
                    pr.y = _parseInt(xy[1]);
                }
                if (wh.length === 2) {
                    pr.width = _parseInt(wh[0]);
                    pr.height = _parseInt(wh[1]);
                }
            }
        }
        if (i < colors.length) pr.color = colors[i];
        if (i < cc.length) pr.countryCode = _parseInt(cc[i]);
        plates.push(pr);
    }
    return plates;
}

function _parseBrackets(raw) {
    const out = [];
    if (!raw) return out;
    let pos = 0;
    while (pos < raw.length) {
        const o = raw.indexOf('[', pos);
        if (o < 0) break;
        const c = raw.indexOf(']', o);
        if (c < 0) break;
        out.push(raw.slice(o + 1, c));
        pos = c + 1;
    }
    return out;
}

function extractObjectDetections(tags) {
    const cls = (tags && tags.ClassifierList) ? tags.ClassifierList : '';
    if (!cls) return [];

    const clsEntries = _parseBrackets(cls);
    const bmcEntries = _parseBrackets((tags && tags.BMCList) || '');

    const detections = [];
    for (let i = 0; i < clsEntries.length; i++) {
        const parts = clsEntries[i].split(',');
        if (parts.length < 6) continue;
        const det = {
            type: _parseInt(parts[0]),
            probability: _parseInt(parts[1]) / 100.0,
            x: _parseInt(parts[2]),
            y: _parseInt(parts[3]),
            width: _parseInt(parts[4]),
            height: _parseInt(parts[5]),
            brand: '', brandProb: 0,
            model: '', modelProb: 0,
            color: '', colorProb: 0,
        };
        if (i < bmcEntries.length) {
            const cleaned = bmcEntries[i].replace(/"/g, '');
            const bp = cleaned.split(',');
            if (bp.length >= 6) {
                det.brand = bp[0];
                det.brandProb = _parseInt(bp[1]) / 100.0;
                det.model = bp[2];
                det.modelProb = _parseInt(bp[3]) / 100.0;
                det.color = bp[4];
                det.colorProb = _parseInt(bp[5]) / 100.0;
            }
        }
        detections.push(det);
    }
    return detections;
}

function parseJpegMetadata(jpeg) {
    const comment = extractJpegComment(jpeg);
    if (!comment) {
        return { comment: '', tags: {}, plates: [], objects: [] };
    }
    const tags = parseJpegCommentTags(comment);
    return {
        comment,
        tags,
        plates: extractPlateRecognitions(tags),
        objects: extractObjectDetections(tags),
    };
}

module.exports = {
    extractJpegComment,
    parseJpegCommentTags,
    extractPlateRecognitions,
    extractObjectDetections,
    parseJpegMetadata,
};
